segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/TableEntryIterator.java

```diff
@@ -0,0 +1,328 @@
+    @GuardedBy("this")
+    private Iterator<Map.Entry<EntryIteratorState, TableEntry>> currentEntry = null;
+    @GuardedBy("this")
+    private long currentBatchOffset;
+    //endregion
+
+    private TableEntryIterator<T> setCurrentBatchOffset(long currentBatchOffset) {
+        this.currentBatchOffset = currentBatchOffset;
+        return this;
+    }
+
+    //region AsyncIterator Implementation
+
+    @Override
+    public CompletableFuture<T> getNext() {
+        // Verify no other call to getNext() is currently executing.
+        return getNextEntry()
+                .thenCompose(entry -> {
+                    if (entry == null) {
+                        return CompletableFuture.completedFuture(null);
+                    } else {
+                        return this.resultConverter.apply(entry);
+                    }
+                });
+    }
+
+    public boolean endOfSegment() {
+        return this.currentBatchOffset >= (startOffset + maxLength);
+    }
+
+    private CompletableFuture<Map.Entry<EntryIteratorState, TableEntry>> getNextEntry() {
+        val entry = getNextEntryFromBatch();
+        if (entry != null) {
+            return CompletableFuture.completedFuture(entry);
+        }
+
+        return fetchNextTableEntriesBatch().thenApply(val -> getNextEntryFromBatch());
+    }
+
+    private synchronized Map.Entry<EntryIteratorState, TableEntry> getNextEntryFromBatch() {
+        if (this.currentEntry != null) {
+            Map.Entry<EntryIteratorState, TableEntry> next = this.currentEntry.next();
+            if (!this.currentEntry.hasNext()) {
+                this.currentEntry = null;
+            }
+            return next;
+        }
+
+        return null;
+    }
+
+    private CompletableFuture<Void> fetchNextTableEntriesBatch() {
+        return toEntries(currentBatchOffset)
+                .thenAccept(entries -> {
+                    if (!entries.isEmpty()) {
+                        this.currentEntry = entries.iterator();
+                    } else {
+                        this.currentEntry = null;
+                    }
+                });
+    }
+
+    private CompletableFuture<List<Map.Entry<EntryIteratorState, TableEntry>>> toEntries(long startOffset) {
+        TimeoutTimer timer = new TimeoutTimer(this.fetchTimeout);
+        int length = Math.min(maxLength, MAX_READ_SIZE);
+
+        if (endOfSegment()) {
+            return CompletableFuture.completedFuture(Collections.emptyList());
+        }
+        ReadResult result = this.segment.read(startOffset, length, timer.getRemaining());
+        return AsyncReadResultProcessor.processAll(result, this.executor, timer.getRemaining())
+                .thenApply(data -> parseEntries(data, startOffset, length));
+    }
+
+    @SneakyThrows(IOException.class)
+    private List<Map.Entry<EntryIteratorState, TableEntry>> parseEntries(BufferView data, long startOffset, int readLength) {
+
+        long maxOffset = startOffset + readLength;
```

Comment:
Tip: if you declare a variable that you do not expect (or want) to change throughout your method, add the `final` modifier or declare it with `val` (which makes it final). This way the compiler will catch any attempt you make to modify it, which can, in turn, prevent bugs from developing.